zcash_client_sqlite/src/for_async/mod.rs

```diff
@@ -0,0 +1,488 @@
+        account: AccountId,
+        anchor_height: BlockHeight,
+    ) -> Result<Amount, Self::Error> {
+        let db = self.clone();
+        async_blocking(move || {
+            let db = db.inner.lock().unwrap();
+            wallet::get_balance_at(&db, account, anchor_height)
+        })
+        .await
+    }
+
+    async fn get_memo(&self, id_note: Self::NoteRef) -> Result<Memo, Self::Error> {
+        let db = self.clone();
+        async_blocking(move || {
+            let db = db.inner.lock().unwrap();
+            match id_note {
+                NoteId::SentNoteId(id_note) => wallet::get_sent_memo(&db, id_note),
+                NoteId::ReceivedNoteId(id_note) => wallet::get_received_memo(&db, id_note),
+            }
+        })
+        .await
+    }
+
+    async fn get_commitment_tree(
+        &self,
+        block_height: BlockHeight,
+    ) -> Result<Option<CommitmentTree<Node>>, Self::Error> {
+        let db = self.clone();
+        async_blocking(move || {
+            let db = db.inner.lock().unwrap();
+            wallet::get_commitment_tree(&db, block_height)
+        })
+        .await
+    }
+
+    #[allow(clippy::type_complexity)]
+    async fn get_witnesses(
+        &self,
+        block_height: BlockHeight,
+    ) -> Result<Vec<(Self::NoteRef, IncrementalWitness<Node>)>, Self::Error> {
+        let db = self.clone();
+        async_blocking(move || {
+            let db = db.inner.lock().unwrap();
+            wallet::get_witnesses(&db, block_height)
+        })
+        .await
+    }
+
+    async fn get_nullifiers(&self) -> Result<Vec<(AccountId, Nullifier)>, Self::Error> {
+        let db = self.clone();
+        async_blocking(move || {
+            let db = db.inner.lock().unwrap();
+            wallet::get_nullifiers(&db)
+        })
+        .await
+    }
+
+    async fn get_spendable_notes(
+        &self,
+        account: AccountId,
+        anchor_height: BlockHeight,
+    ) -> Result<Vec<SpendableNote>, Self::Error> {
+        let db = self.clone();
+        async_blocking(move || {
+            let db = db.inner.lock().unwrap();
+            wallet::transact::get_spendable_notes(&db, account, anchor_height)
+        })
+        .await
+    }
+
+    async fn select_spendable_notes(
+        &self,
+        account: AccountId,
+        target_value: Amount,
+        anchor_height: BlockHeight,
+    ) -> Result<Vec<SpendableNote>, Self::Error> {
+        let db = self.clone();
+        async_blocking(move || {
+            let db = db.inner.lock().unwrap();
+            wallet::transact::select_spendable_notes(&db, account, target_value, anchor_height)
+        })
+        .await
+    }
+}
+
+#[derive(Clone)]
+pub struct DataConnStmtCacheAsync<P> {
+    wallet_db: WalletDbAsync<P>,
+}
+
+impl<P: consensus::Parameters> DataConnStmtCacheAsync<P> {
+    fn transactionally<F, A>(&mut self, f: F) -> Result<A, SqliteClientError>
+    where
+        F: FnOnce(&mut Self) -> Result<A, SqliteClientError>,
```

Comment:
`mut` is not required here because `DataConnStmtCacheAsync` is eventually wrapper over `Arc`. I think you may even use `self` here (not reference), which likely allows putting it in `async_blocking`.

Reply:
I can make this work if you it's okay to refactor `PrunedBlock`, `ReceivedTransaction` and `SentTransaction` and the usages..i.e by removing lifetimes from the struct and wrapping them in `Mutex` since it's used atomically. <https://github.com/KomodoPlatform/librustzcash/pull/8#discussion_r1344688376> @artemii235

Reply:
found a better solution..so done

Reply:
I assume one of the goals is to avoid making changes in original lib code. As a workaround, we can use [block_in_place](https://docs.rs/tokio/latest/tokio/task/fn.block_in_place.html). Though, it requires `rt-multi-thread` feature. Is it acceptable to enable it?

Reply:
yeah we can use..done thanks